### DISCUSSION TITLE
Semi-manually generating QSTRs for a user MicroPython library?

### DISCUSSION THREAD

**Discussion by WebDust21** [micropython/micropython#9892] [Core Development]
Big question...resulting from big frustration...resulting from days of banging my head against every which wall.........

It appears to me that an "official" MicroPython build automatically generates the QSTR files at compile-time, and this is a non-issue.
Or that "most" people just use existing ports that "just work" on a very select number of hardware options.  Or that "most" people don't try to push the limits to realize an intended vision.

While I could (and will!) be very specific as to what exactly I'm trying to do, I don't feel that this is a specific question--but rather one that has the potential to significantly ease the difficulty of porting MicroPython to ever more obscure boards/frameworks/compilers, etc.

So, first, my system setup/toolchains

- Ubuntu 22 x64.  No Windoze for me....
- I am using RT-Thread 5.0.0.  Which has MicroPython v1.13-148-ged7ddd4 in it.  (Not the world's newest, but that shouldn't be a problem.)
- RT-Thread is not built with Make or CMake.  Their project manager of choice is Scons.  (This is probably the biggest issue I'm facing.)
- RT-Thread also has LVGL 8.3.1 as an available library.  Works great as a graphical display manager.
- ...you guessed it: RT-Thread does NOT have LVGL bindings available for MicroPython.  Shocker, right?

I've been scouring the seemingly empty Web search results for the past couple of days trying to figure this stuff out.  A lot of bugfixes required in RT-Thread to get things to compile (things like "return NULL;" in a stubbed-out [for BSP reasons] MicroPython function that's defined as "mp_uint_t"), but after the necessary corrections have been made, MicroPython works without any issues.

Next was generating the MicroPython bindings library, which interfaced between MicroPython endpoints and the LVGL C endpoints.  A repository here seemed to provide a lot of help in that regard: https://github.com/lvgl/lv_binding_micropython/tree/master/gen
...well, after fixing numerous errors in it resulting from updates to "pycparser"...which would be my fault for cheating and installing "pycparser" via PIP3 instead of by Git download from the specified age...

Armed with the generated output file, I figured it should be easy to follow the rest of the instructions and patch the various parts all into the MicroPython ecosystem.

15,000 compile errors later (multiple errors for each line being complained about in the generated C library), I realized that something crucial was missing.

And that "something" appears to be an updated QSTR list in "/qstrdefs.generated.h".

Here then is the problem: The RT-Thread MicroPython port has never heard of automatic QSTR generation.
And the other problem: I'm horrible at understanding Makefiles.  And I've only heard of Scons since I started using RT-Thread.  So trying to translate between the 2...sometimes I wish I could snap my fingers and *poof* it'd be fixed...

So my question: How do I manually generate the QSTRdefs?  I'm talking command prompt-level stuff...not a fully automated "push this button and say bingo" solution that most people seem to expect.

I have the generated bindings library (generated from my specific system setup), "lv_mpy.c".  A cool 1.2MB.
I've noticed five Python files in the "micropython/py" folder, which seem that they are responsible and/or the tools I need to generate the necessary files.

- makecompresseddata.py
- makemoduledefs.py
- makeqstrdata.py
- makeqstrdefs.py
- makeversionhdr.py

Can someone provide the necessary command-line-level utilization steps/instructions for these scripts?  Or even a copy of (verbose) compiler verbiage of the QSTR generation part of a normal compile of MicroPython?  (I could definitely work with that.)
and yes, I do understand there's "best case practice", and then there "desperate case practice"!  I don't mind doing this manually and/or manually regenerating the QSTR files specifically for my "lv_mpy.c" bindings file if/as necessary.

Obviously, once I understand how to manually do this, I will be able to figure out how to automate it with the Scons build environment.  Or just wrap all of it into a bigger PY file...

EDIT: by "verbose compiler verbiage", I mean where the output includes the "make*******.py" tool command lines and all arguments, as well as any output from said tools.

**Comment by jimmo** [comment]
Have you seen https://docs.micropython.org/en/latest/develop/qstr.html which at least provides a high-level overview of the steps involved in QSTR generation (might help you understand the makefiles).

Realistically though, supporting this process across more build systems (we also do cmake, which is also a hassle) isn't really scalable. The same is true for IDE build systems (e.g. STM32 Cube IDE) and lots of others.

Instead I think the better approach is to build MicroPython using the existing make system, and then integrate the output of that into your build system.

See https://github.com/micropython/micropython/pull/9529 for the first step towards this -- the idea eventually is that you run the MicroPython build on micropython itself as well as your bindings, and the the output of this can just be included as .c (and .h) files into your SCons or whatever.

**Reply by WebDust21** [reply]
Yes, I've been through that doc several times.  Like you said, it provides a "high-level overview"--the general concept without enough specifics to actually be able to utilize it (unless you already know how it works...in which case you wouldn't need the doc).

I'm not asking for MicroPython to "support more build systems"--at all.  That would actually defeat the purpose that I'm trying to attain here ;-).

What I'm trying to understand is how the QSTR generation process works _in the current official MicroPython setup_.  Because concepts are pretty much always platform-independent.  How the QSTR Python scripts are _actually_ utilized...not "oh, the output of this Python script goes into that Python script, which does this, and then goes into that Python script (like the qstr.htm doc).

**Reply by jimmo** [reply]
> I'm not asking for MicroPython to "support more build systems"--at all. That would actually defeat the purpose that I'm trying to attain here ;-).

Sorry I phrased that poorly. What I meant was that having people add support for additional build systems (in your case, SCons) is likely a difficult task, and it would be better if we made it easier to use the "official" (make-based) mechanism used in the MicroPython build from third-party systems.

> How the QSTR Python scripts are actually utilized...not "oh, the output of this Python script goes into that Python script, which does this, and then goes into that Python script (like the qstr.htm doc).

One thing that might help you is to run `make V=1` on one of the existing ports, which will print out the command line for each command.

e.g.

```
cd ports/stm32
make BOARD=PYBV11 V=1 build-PYBV11/genhdr/qstrdefs.generated.h
```

which will show the steps up to the completion of the QSTR generation.

**Reply by WebDust21** [reply]
> Have you seen https://docs.micropython.org/en/latest/develop/qstr.html which at least provides a high-level overview of the steps involved in QSTR generation (might help you understand the makefiles).
> 
> Realistically though, supporting this process across more build systems (we also do cmake, which is also a hassle) isn't really scalable. The same is true for IDE build systems (e.g. STM32 Cube IDE) and lots of others.
> 
> Instead I think the better approach is to build MicroPython using the existing make system, and then integrate the output of that into your build system.
> 
> See #9529 for the first step towards this -- the idea eventually is that you run the MicroPython build on micropython itself as well as your bindings, and the the output of this can just be included as .c (and .h) files into your SCons or whatever.

#9529 is not based around generating QSTRdefs, but appears to be more of a means to compile and embed Python scripts into the actual MicroPython build.

I've tried getting the QSTRdefs generated in the MicroPython UNIX build and then copying the generated files over--but of course this is an exercise in exasperation due to the (necessary) MicroPython library differences in MPY UNIX port vs the RT-Thread port.  For the most part it works--but then starts failing on every single difference where MPY UNIX doesn't have a particular library that RT-Thread does.
at the moment, that's:

-  "mpirq.c"...took several hours to _finally_ get that one in MPY UNIX.  Turns out the QSTRdefs generators are terrible at identifying project changes.  As a last resort, I deleted the entire "build-standard" folder, and *poof* MPY UNIX mysteriously compiled without errors (it would fail due to missing QSTRdefs in "mpirq.c" which I'd added to the build list, because the QSTRdefs generators were failing to recognize the changes).  Oh, and the resulting "qstrdefs.generated.h" file was all of 520 bytes bigger...neatly adding on the 3-4 QSTRs for "mpirq.c".  Copied "qstrdefs.generated.h" over to RT-Thread, and it was compiling well until it hit...
- "machine_pin.c".  This one's a real doozy of a problem, because it's highly dependent on the platform MPY is being built on.  Looking through the MPY ports, this file ranges from 12kb to 25kb depending on the port.  And I can pretty well guarantee that that's going to result in different QSTRdef requirements.  Oh, and did I mention that of the MPY ports utilizing "machine_pin.c"...NONE of them utilize "mpirq.c"?  That's just 2 files...there's almost unquestionably more.


#5964 is not much help either, as it has to do with "adding an external library" to MicroPython, "adding in" a minimal QSTRdef file for the external library alone.  Like #9259, it also does not generate project-level QSTRdefs.

**Reply by robert-hh** [reply]
> NONE of them utilize "mpirq.c"? 

That's not true. Quite a few ports use mpirq.c, like stm32, renesas-ra, mimxrt, rp2, c3200, samd, and zephyr.

**Reply by WebDust21** [reply]
That's not what I wrote: of the ports utilizing "machine_pin.c" NONE of those _also_ utilize "mpirq.c".  From what I could find anyway.

At least in the MicroPython ("lv_micropython") I'm trying to work with, the ONLY port that has "mpirq.c" in the port folder, is c3200.  (It's possible that others might pull it from the "shared/runtime" folder.)  Eight ports have "machine_pin.c", but c3200 is not one of them.

**Reply by robert-hh** [reply]
> It's possible that others might pull it from the "shared/runtime" folder.

That's where it is. Instead of having copies of the same file in every port folder, the aim is to have the same file for the same task and keep it at a single place.

**Comment by WebDust21** [comment]
One potential issue I see with trying to patch together multiple "qstrdef" files, is the use of the word "index" 5 times on the linked qstr.html page.
That kinda points to the qstrdefs needing to be generated "in situ", or at least with the exact same libraries in the generation location as in the target location.  Because if a sequential index doesn't line up (or if there are duplicate index numbers) as a result of patching together several qstrdef files, the whole show comes to a screeching halt in runtime.

**Comment by WebDust21** [comment]
So I managed to get an example of QSTRdefs generation from the "lv_micropython" port.  Hacked and cobbled....over a week of exasperatingly frustrating days I managed to get a QSTR file built with the necessary QSTRs that DID work in the RT-Thread port.  (And I do mean _exasperating_.)

Anyway, looking at the verbose build output log (650KB!), I'm more confused than before.

ALL I need for an output is the "qstrdefs.generated.h" output file.  Which contains NOTHING except a bazillion QSTR definition lines (OK, well, 2,989).

A QSTR definition example is:
```
QDEF1(MP_QSTR_anim_timeline_del, 88, 17, "anim_timeline_del")
```

- "QDEFx" vs "QDEF" (with no number) ...the difference there appears to be a recent MicroPython change, present in the 1.19 of "lv_micropython", but not in 1.13 of "RT-Thread".  (FWIW I managed to hack the newer "qstr.h"/"qstr.c" into RT-Thread to utilize the resulting QSTRdefs.h file.)  Anyone want to shine light on what the difference between "QSTR0" and "QSTR1" is...unless it's just a different dictionary table?
- The first argument appears to be the reference to the QSTR utilized in the C source files.
- Second argument...some weird sort of index that goes from 0-255 very loosely from the beginning to the end of the "qstrdefs.generated.h" file.  Understanding what this is would be very helpful--as it's DEFINITELY not a "unique index" by any metric.
- Third argument: string length, in bytes.
- Fourth argument: the actual string.  It appears at a glance that this string COMES from the QSTR title itself.  Which is extremely simple to do.

This looks relatively straightforward.


But the QSTR generation process starts with:
```
python3 ../../py/makeqstrdefs.py pp gcc -E output build-standard/genhdr/qstr.i.last [........lots of arguments, list of basically every single MicroPython file--understood for scanning to find "QSTR" defs]
```
This generates a truly ginormous 150MB file, containing hundreds of duplicate lines.  It appears to be entirely definitions, structs, and variables.  What in the world is this used for?

This call also appears to create a huge array of folders with ".qstr" files for every source file.  Each of these files contain a list of very succinct QSTR definitions found in the source, i.e.: 
```
Q(LV_TEXT_DECOR)
```
Which this can be very easily utilized to generate a QSTRdef--at least every single part of it except for the one index argument.

I can understand a lot of the GCC steps and arguments as necessary for interpreting #ifdef conditional defines and such like that.  Personally, I don't care/need conditional QSTR generation.  That can be omitted.

The next QSTRdef generation step is:
```
python3 ../../py/makeqstrdefs.py split qstr build-standard/genhdr/qstr.i.last build-standard/genhdr/qstr _
touch build-standard/genhdr/qstr.split
```
This appears to generate "qstr.split" -- which at the end of the build process is a 0-byte empty file.  Useless step?

Next step:
```
python3 ../../py/makeqstrdefs.py cat qstr _ build-standard/genhdr/qstr build-standard/genhdr/qstrdefs.collected.h
```
This appears to go through all of the QSTR files in the "build-standard/genhdr/qstr" directory (created by the very first line?) and concatenate + alphabetically sort them into "qstrdefs.collected.h".  This resulting file has thousands of duplicate entries (sometimes 10x duplicates of a single entry).

Next instruction:
```
cat ../../py/qstrdefs.h qstrdefsport.h build-standard/genhdr/qstrdefs.collected.h | sed 's/^Q(.*)/"&"/' | gcc -E -I../../lib/lv_bindings -I../../lib/berkeley-db-1.xx/PORT/include -I. -I../.. -Ibuild-standard -I../../shared/readline -Wall -Werror -Wextra -Wno-unused-parameter -Wpointer-arith -Wdouble-promotion -Wfloat-conversion -std=gnu99 -DUNIX   -Wno-unused-function -DFFCONF_H=\"lib/oofatfs/ffconf.h\" -DMICROPY_PY_USSL=1 -DMICROPY_SSL_AXTLS=1 -I../../lib/axtls/ssl -I../../lib/axtls/crypto -I../../extmod/axtls-include -DMICROPY_PY_BTREE=1 -DMICROPY_USE_READLINE=1 -DMICROPY_PY_TERMIOS=1 -DMICROPY_PY_SOCKET=1 -DMICROPY_PY_THREAD=1 -DMICROPY_PY_THREAD_GIL=0  -DMICROPY_PY_FFI=1 -Os -DNDEBUG -fdata-sections -ffunction-sections -Ivariants/standard  -g -U _FORTIFY_SOURCE -DMICROPY_QSTR_EXTRA_POOL=mp_qstr_frozen_const_pool -DMICROPY_MODULE_FROZEN_MPY -DMPZ_DIG_SIZE=16  -DMICROPY_MODULE_FROZEN_STR - | sed 's/^\"\(Q(.*)\)\"/\1/' > build-standard/genhdr/qstrdefs.preprocessed.h
```
This generates "qstrdefs.preprocessed.h".  This file appears to be some sort of SEVERELY condensed "qstr.i.last" followed by a direct dump of "qstrdefs.collected.h", duplicates and all.  I still do not understand in the slightest what the purpose/need of "qstr.i.last" is in the generation of QSTRdefs.

Closely related to the above line is:
```
python3 ../../py/makeqstrdata.py build-standard/genhdr/qstrdefs.preprocessed.h > build-standard/genhdr/qstrdefs.generated.h
```
which throws out all of the definitions, and pulls the actual QSTRdefs together into the only necessary output file.


Unless there's some black magic with the number in the 2nd argument in a QSTRdef, I don't see why this can't be DRASTICALLY simplified * 100 into a VERY simple QSTR generation routine:

1. Scan all MicroPython source files for "QSTR(" entries.  Add each entry to a Python dictionary--which will automatically throw out duplicates right here.  It can also very easily alphabetically sort.
2. Fill out the QSTR entries into the "qstrdefs.generated.h" file, with the string & length arguments -- and whatever that 2nd argument is.

As I said, I don't care about conditional project #includes that determine what MicroPython modules are/aren't included--which will DRASTICALLY simplify the process, as the entire project's #include and #define structures do not need to be interpreted by the QSTR generation process.  Heck, the RT-Thread stock "qstrdefs.generated.h" file just contains every single QSTR for every module--whether used or not.


If all of the above is condensed into a succinct question, it would be:
What is the significance/purpose of the 2nd argument in a QSTR definition in "qstrdefs.generated.h"?

**Reply by WebDust21** [reply]
Think I answered my own questions...and yes, there's no reason why this hugely overcomplicated routine can't be broken down into a VERY simple process.......

"QDEF0" is simply built-in QSTRs that are autogenerated regardless of the source (at least for MicroPython 1.19; in 1.13, there are only ever "QDEF" entries, no number.)
"QDEF1" is everything else found in the source files.

Here's the routine in "makeqstrdata.py" that does the final outputting:
```
def print_qstr_data(qstrs):
    # print out the starter of the generated C header file
    print("// This file was automatically generated by makeqstrdata.py")
    print("")

    # add NULL qstr with no hash or data
    print('QDEF0(MP_QSTRnull, 0, 0, "")')

    # split qstr values into two pools. static consts first.
    q0_values = [q for q in qstrs.values() if q.order < 0]
    q1_values = [q for q in qstrs.values() if q.order >= 0]

    # go through each qstr in pool 0 and print it out. pool0 has special sort.
    for q in sorted(q0_values, key=lambda x: x.order):
        print('QDEF0(MP_QSTR_%s, %d, %d, "%s")' % (q.ident, q.qhash, q.qlen, q.qdata))

    # go through each qstr in pool 1 and print it out. pool1 is regularly sorted.
    for q in sorted(q1_values, key=lambda x: (x.qhash, x.qlen)):
        print('QDEF1(MP_QSTR_%s, %d, %d, "%s")' % (q.ident, q.qhash, q.qlen, q.qdata))
```
Second argument?  "q.qhash"

That's a class function of the qstr class, which is defined as follows:
```
@property
    def qhash(self):
        return compute_hash(self.qbytes, Qstr.cfg_bytes_hash)
```

And "compute_hash" is this little routine:
```
def compute_hash(qstr, bytes_hash):
    hash = 5381
    for b in qstr:
        hash = (hash * 33) ^ b
    # Make sure that valid hash is never zero, zero means "hash not computed"
    return (hash & ((1 << (8 * bytes_hash)) - 1)) or 1
```

So there is absolutely zero use for the 135MB "qstr.i.last" file...that is, if we don't care about compile-time configuration to include/remove various MicroPython libraries.

"Q.ident" is simply an escaping method to make valid C code out of symbols...that's easy.  ("ident = qstr_escape(qstr)")
```
def qstr_escape(qst):
    def esc_char(m):
        c = ord(m.group(0))
        try:
            name = codepoint2name[c]
        except KeyError:
            name = "0x%02x" % c
        return "_" + name + "_"

    return re.sub(r"[^A-Za-z0-9_]", esc_char, qst)
```

.............I'm eating this snake one byte at a time.  And maybe making the MPY world a bit better for other hapless humans like myself who almost got eaten alive and gave up ;-)

**Comment by Biancaa-R** [comment]
![image](https://github.com/user-attachments/assets/976081af-3d45-4abb-bf3b-b5000b20bb15)
My qstrdefs.generated.h is fine and it is generated , but still I get an error at   assert(*q <pool ->len);  in qstr.c file in py folder when I run it in my FPGA . It is in the find_qstr function. I am really dying with this issue as I have a tight deadline

**Reply by Biancaa-R** [reply]
OK so I realized something , I work in a baremetal environment so a lot of POSIX commands , commands that triggers linux functions will act erroneously.. So the assert function tried calling a higher mode . It resulted in a ECALL from Mmode in my chip (It makes no sense as its already in the highest priority.) Had to change the assert statements in qstr.c ,now it works ..  (I am setting up micropython for a RISCV based chip btw.)

**Comment by Biancaa-R** [comment]
Does the micropython versions post 2023 have a build/genhdr/qstrdefs.generated.c file ? There is a .h file I know ,what about the .c?

OK now I know the answer: No , post 2023 ish , there are only qstrdefs.generated.h file and the qstr.c traverses through the saved Qstrings in qstrdefs....h to find the qstrings and map them .

**Comment by Biancaa-R** [comment]
There is another issue 
![image(2)](https://github.com/user-attachments/assets/b8803cd6-4974-44d2-ad77-2bdd7a416258)
I am unable to use in built functions even when the qstrings are generated the mp_rom connecting functions are defined in modbuiltins.c and all the right macros are enabled ,even then there is a issue !

```
`#define MICROPY_PY_IO  (1)
#define MICROPY_PY_BUILTINS_INPUT (1)
#define MICROPY_PY_BUILTINS_OPEN (1)
#define MICROPY_PY_SYS_STDFILES  (1)
#include <stdint.h>

// options to control how MicroPython is built

// Use the minimal starting configuration (disables all optional features).
//#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_MINIMUM)
#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_CORE_FEATURES)
// You can disable the built-in MicroPython compiler by setting the following
// config option to 0.  If you do this then you won't get a REPL prompt, but you
// will still be able to execute pre-compiled scripts, compiled with mpy-cross.
#define MICROPY_PY_SYS           (1)
#define MICROPY_PY_ARRAY         (1)
#define MICROPY_PY_COLLECTIONS   (1)
#define MICROPY_PY_IO            (0)
#define MICROPY_PY_IO_FILEIO      (0)
#define MICROPY_PY_STRUCT        (1)
#define MICROPY_PY_GC            (1)
#define MICROPY_PY_MICROPYTHON   (1)
#define MICROPY_ENABLE_COMPILER     (1)

#define MICROPY_QSTR_EXTRA_POOL           mp_qstr_frozen_const_pool
//#define MICROPY_QSTR_EXTRA_POOL           NULL
//defined in frozen_mpy.c file in ports
#define MICROPY_ENABLE_GC                 (1)
#define MICROPY_HELPER_REPL               (1)
//#define MICROPY_MODULE_FROZEN_MPY         (1)
#define MICROPY_ENABLE_EXTERNAL_IMPORT    (1)
#define MICROPY_READER_POSIX               (0)
#define MICROPY_ENABLE_IMPORT_STAT         (0)
#define MICROPY_ALLOC_PATH_MAX            (256)
#define MICROPY_PY_MACHINE_UART            (1)

#define MICROPY_PY_MACHINE_UART_INCLUDEFILE "./uart.h"
// Use the minimum headroom in the chunk allocator for parse nodes.
#define MICROPY_ALLOC_PARSE_CHUNK_INIT    (16)
#define MICROPY_READER_VFS         (1)  // disables mp_lexer_new_from_file
#define MICROPY_PY_BUILTINS_OPEN    (0)
#define MICROPY_PY_BUILTINS_INPUT      (1)
#define MICROPY_PY_BUILTINS_PRINT      (1)

//#define MICROPY_PY_BUILTINS_OPEN 1
#define MICROPY_ENABLE_FILE_IMPORT (1)  // disables mp_import_stat
#define MICROPY_MODULE_FROZEN_MPY  (0)
#define MICROPY_MODULE_FROZEN (0) //Have to disable this along witb the mpy file also
#define MICROPY_MODULE_FROZEN_STR  (0)
// type definitions for the specific machine
//#define MICROPY_ENABLE_COMPILER         (0)
#define MICROPY_MODULE_WEAK_LINKS       (0)
#define MICROPY_MODULE_BUILTIN_INIT      (1)
//For thye support of the filw sytems
#define MICROPY_VFS               (0)
#define MICROPY_VFS_FAT           (0)   // For FAT filesystem support
//#define MICROPY_MODULE_FROZEN_MPY       (0) //--As we are not using frozen mpy files but py files directly into the build system
#define MICROPY_PY_BUILTINS_HELP       (1)
#define MP_STATE_PORT_ROOT_POINTERS \
    const char *readline_hist[MICROPY_READLINE_HISTORY_SIZE];


typedef intptr_t mp_int_t; // must be pointer size
typedef uintptr_t mp_uint_t; // must be pointer size
typedef long mp_off_t;
//#define MP_STATE_PORT_VM \
    //void *uart0_rxbuf;
//we have to define uart0_rxbuf for future use
// We need to provide a declaration/definition of alloca()
#include <alloca.h>

// #define MICROPY_PORT_BUILTINS \
//    { MP_ROM_QSTR(MP_QSTR_open), MP_ROM_PTR(&mp_builtin_open_obj) },

#define MICROPY_HW_BOARD_NAME "mindgrove-secure-iot"
#define MICROPY_HW_MCU_NAME "secure-iot"
// #define MP_STATE_PORT_ROOT_POINTERS \
// mp_obj_t uart0_rxbuf;
//Using a custom buffer
//Need to declare uart0_rxbuf in port specific mp_state_vm_t

#if defined(__linux__) || defined(__APPLE__)
#define MICROPY_MIN_USE_STDOUT (1)
#define MICROPY_HEAP_SIZE      (25600) // heap size 25 kilobytes
#endif

#ifdef __thumb__
#define MICROPY_MIN_USE_CORTEX_CPU (1)
#define MICROPY_MIN_USE_STM32_MCU (1)
#define MICROPY_HEAP_SIZE      (2048) // heap size 2 kilobytes
#endif

#define MP_STATE_PORT MP_STATE_VM

```

` this is my mpconfigport.h file

**Comment by Biancaa-R** [comment]
![image](https://github.com/user-attachments/assets/95cd59e2-235b-4edc-953d-1200fedd9082)
Pretty children !! I finally after dayys of suffering found out what the issue was :
( Disclaimer : Some of these mentioned methods are only applicable for riscv based system and that is the architecture I am working on.)

1. Ecall !! -> So I am using a baremetal environment where posix commands are not supported <3 ,so Internally during the linking when reading the qstring values ,ECall was triggered ... (Weird how it was not triggered in some of my printfs.) exactly in a fstat statement.
`this code is triggering a posix function--> ecall 000000009003f4ba <_fstat>:
    9003f4ba:	7135                	addi	sp,sp,-160
    9003f4bc:	e526                	sd	s1,136(sp)
    9003f4be:	ed06                	sd	ra,152(sp)
    9003f4c0:	84ae                	mv	s1,a1
    9003f4c2:	e922                	sd	s0,144(sp)
    9003f4c4:	05000893          	li	a7,80
    9003f4c8:	858a                	mv	a1,sp
    9003f4ca:	00000073          	ecall
    9003f4ce:	842a                	mv	s0,a0
    9003f4d0:	00054e63          	bltz	a0,9003f4ec <_fstat+0x32>
    9003f4d4:	0005041b          	sext.w	s0,a0
    9003f4d8:	8526                	mv	a0,s1
    9003f4da:	858a                	mv	a1,sp
    9003f4dc:	134000ef          	jal	ra,9003f610 <_conv_stat>
    9003f4e0:	60ea                	ld	ra,152(sp)
    9003f4e2:	8522                	mv	a0,s0
    9003f4e4:	644a                	ld	s0,144(sp)
    9003f4e6:	64aa                	ld	s1,136(sp)
    9003f4e8:	610d                	addi	sp,sp,160
    9003f4ea:	8082                	ret
    9003f4ec:	4080043b          	negw	s0,s0
    9003f4f0:	ce9ef0ef          	jal	ra,9002f1d8 <__errno>
    9003f4f4:	c100                	sw	s0,0(a0)
    9003f4f6:	547d                	li	s0,-1
    9003f4f8:	b7c5                	j	9003f4d8 <_fstat+0x1e>`
2. so add this in your makefile `CFLAGS += -fno-builtin-printf
CFLAGS += -fno-builtin-memcpy
CFLAGS += -fno-builtin-malloc`
3. #define MICROPY_QSTR_BYTES_IN_HASH (0) --> If you are not using micropython in frozen mode and saving frozen qstrings its very advisable to disable it .
4. Note : The generated qstrings were not properly copied in the linking process -> That is what this code addresses .
* You can see __modules_start and end : I had used a separate section called modules to store the qstrings for use.
```
extern const struct _mp_obj_module_t *__modules_start;
extern const struct _mp_obj_module_t *__modules_code_end;

void mp_module_init_from_section(void) {
    const struct _mp_obj_module_t **mod;

    for (mod = &__modules_start; mod < &__modules_code_end; ++mod) {
        mp_obj_t mod_obj = MP_OBJ_FROM_PTR(*mod);
        //Iterates over statically linked modules placed in the .modules section
        mp_map_elem_t *name_elem = mp_map_lookup(
            &((mp_obj_module_t *)(*mod))->globals->map,
            MP_OBJ_NEW_QSTR(MP_QSTR___name__),
            MP_MAP_LOOKUP
            //If it's a valid QSTR, stores the module into sys.modules
        );
        printf("MOD = %p\n", mod); //to show the value pointed (address in the modules section)
        printf("*MOD = %p\n", *mod);
        if (*mod == NULL) continue; 
        printf("MOD->globals = %p\n", (*mod)->globals);
        if (name_elem && mp_obj_is_qstr(name_elem->value)) {
            qstr mod_qstr = mp_obj_str_get_qstr(name_elem->value);
            //Initialized (qstr_init())
            //Contain MP_QSTR_mymodule, MP_QSTR___name__,
            mp_obj_dict_store(MP_OBJ_FROM_PTR(&MP_STATE_VM(mp_loaded_modules_dict)),
            //// dictionary with loaded modules (may be exposed as sys.modules)
                              MP_OBJ_NEW_QSTR(mod_qstr), mod_obj);
        }
    }
}

```


If issue with the macro for registering qstrings use mine (change according to linker script.) Made by me with love :)

6. Check that your root_pointers.h values point to VM not PORT ! --> Silly but deadly :)
I mean this "#define MP_STATE_VM(x) (mp_state_ctx.vm.x)" in mpstate.h

now builtin functions ,predefined qstrings and functions from other modules works :)
